Add named entities widget

paltman opened the pull request.

paltman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [As a user, I can see a list of named entities in the passage](https://trello.com/c/WPEGWVfP/102-as-a-user-i-can-see-a-list-of-named-entities-in-the-passage)

paltman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [As a user, I can select a named entity in the passage and update the named entities widget to see more detail](https://trello.com/c/oDQlPPBA/103-as-a-user-i-can-select-a-named-entity-in-the-passage-and-update-the-named-entities-widget-to-see-more-detail)

paltman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [As a user, when I select a named entity occurrence in the text, all of the other occurrences are highlighted, and the selection in the widget is updated](https://trello.com/c/GBug4YiK/104-as-a-user-when-i-select-a-named-entity-occurrence-in-the-text-all-of-the-other-occurrences-are-highlighted-and-the-selection-in)

jacobwegner commented on src/widgets/NamedEntitiesWidget/NamedEntitiesWidget.vue:
```diff
@@ -0,0 +1,128 @@
+<template>
+  <div class="named-entities">
```
We're missing a way to clear the selection from the widget:

From 
![](https://github.trello.services/images/mini-trello-icon.png) [As a user, I can see a list of named entities in the passage](https://trello.com/c/WPEGWVfP)
> Have a way to clear the selection in the widget (to see all the entities)

![image](https://user-images.githubusercontent.com/629062/82500377-e089ab00-9ab8-11ea-9aea-a14271946e54.png)

jacobwegner replied:
de-selection is also referenced in ![](https://github.trello.services/images/mini-trello-icon.png)[As a user, I can select a named entity in the passage and update the named entities widget to see more detail](https://trello.com/c/oDQlPPBA).

A "clear" option in the widget would be nice, but in addition I'm open to something with a modifier key like we have in the "highlight" mode on [scaife.perseus.org](https://scaife.perseus.org/reader/urn:cts:greekLit:tlg0012.tlg001.perseus-grc2:1.1-1.7/?highlight=%40%CE%A0%CE%B7%CE%BB%CE%B7%CF%8A%CE%AC%CE%B4%CE%B5%CF%89%5B1%5D).

(I'm _not_ proposing a highlight mode, but rather a way to clear the selection without having to click a "clear" button in the widget).

paltman replied:
To conserve space in the widget, I just made it a toggle so if you click on the entity again it unselects it.  I could put a little X in top right of the "well" if that is more intuitive.

jacobwegner replied:
The `X` helps; thanks.

jacobwegner commented:
Again, due to a lagging bug with `scaife-widgets`, this is a good preview URL:

https://deploy-preview-10--explorehomer-dev.netlify.app/reader?urn=urn%3Acts%3AgreekLit%3Atlg0012.tlg001.perseus-grc2%3A1.1-1.7

jacobwegner commented on src/reader/components/ReaderView.vue:
```diff
@@ -50,6 +51,7 @@
           WordListWidget,
           TokenAnnotationWidget,
           NewAlexandriaWidget,
+          NamedEntitiesWidget,
```
I know there was discussion on this previously, but I think it might be good to have the annotation widget and named entities widgets higher in the list; higher position means it is more likely the widgets will be in the "default" viewport for a user and it will be clearer that selecting a token updates the widget contents (unlike `WordListWidget` and `NewAlexandriaWidget` which do not respond to token selections)

jacobwegner commented:
@paltman I've just updated `develop` with the latest `scaife-widgets`; if you pull those changes in, that should resolve the console error.

paltman commented:
@jacobwegner I think this is ready for another review.  I've addressed the two asks by @jtauber (I think I did them right) and cleaned up other UI issues I remember us talking about.

jacobwegner commented:
@paltman thanks.

re: #1 above, the 'selection' green works great once you've selected an entity, but what about styling the unselected tokens when in named entity mode?

> [We style tokens that have references to named entities](https://trello.com/c/GBug4YiK)

(with the idea being that when in named entity mode, you can see where all of the entities are in the text at a glance, and then make selections to see more info in the widget).

jacobwegner commented:
re #2 above, I don't think this has been resolved yet.

Below, when I select the last token on line 7 in "default" mode, the other named entity instances are also selected

<img src="https://cl.ly/477b5a21359c/Screen%20Recording%202020-06-01%20at%2012.14%20PM.gif"/>

paltman commented:
@jacobwegner ok, ready for another review :)

paltman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Token styling improvements](https://trello.com/c/6FsyI9fe/115-token-styling-improvements)

paltman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Ensure consistent size for input fields](https://trello.com/c/EClGZqsX/114-ensure-consistent-size-for-input-fields)

paltman commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Ensure consistent typefaces](https://trello.com/c/z0TcOQZJ/112-ensure-consistent-typefaces)

jacobwegner commented on src/widgets/NamedEntitiesWidget/NamedEntitiesWidget.vue:
```diff
@@ -0,0 +1,132 @@
+          this.filteredEntities = this.entities;
+        },
+      },
+      // selectedEntities: {
```
let's drop the commented out block

jacobwegner commented on src/widgets/NamedEntitiesWidget/NamedEntitiesWidget.vue:
```diff
@@ -0,0 +1,132 @@
+      selectedToken() {
+        return this.$store.state[MODULE_NS].selectedToken;
+      },
+      // selectedEntities() {
```
drop the commented out block

jacobwegner commented on src/mutations.js:
```diff
@@ -3,15 +3,27 @@ import {
+  },
   [SET_DISPLAY_MODE_INTERLINEAR]: state => {
-    state.interlinear = true;
+    state.displayMode = 'interlinear';
```
not a blocker for merge, but as we continue to add more display modes, we might eventually want the equivalent of a Django `ChoiceField` (constants and human-readable label)
